akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventSourcedBehaviorImpl.scala

```diff
@@ -193,8 +185,25 @@ private[akka] final case class EventSourcedBehaviorImpl[Command, Event, State](
     copy(recovery = Recovery(selection.toUntyped))
   }
 
-  override def withRetention(criteria: RetentionCriteria): EventSourcedBehavior[Command, Event, State] =
-    copy(retention = criteria)
+  override def snapshotWhen(predicate: (State, Event, Long) => Boolean): EventSourcedBehavior[Command, Event, State] = {
+    if (retention != RetentionCriteria.disabled)
+      throw new IllegalArgumentException(s"RetentionCriteria [$retention] can't be combined with `snapshotWhen`.")
+    copy(snapshotWhen = predicate)
+  }
+
+  override def withRetention(criteria: RetentionCriteria): EventSourcedBehavior[Command, Event, State] = {
+    criteria match {
+      case DisabledRetentionCriteria => copy(retention = criteria)
+      case s: SnapshotRetentionCriteriaImpl =>
+        if (retention != RetentionCriteria.disabled)
+          throw new IllegalArgumentException(s"RetentionCriteria [$retention] has already been defined.")
+        if (snapshotWhen ne ConstantFun.scalaAnyThreeToFalse)
+          throw new IllegalArgumentException(
+            s"`snapshotWhen` has already been defined, and can't be combined " +
+            s"with RetentionCriteria [$criteria].")
```

Comment:
This might not be great. I see no good way to combine the `snapshotWhen` predicate with `retentionCriteria`. One thought was to move `snapshotWhen` into a separate `RetentionCriteria` but I think that can be problematic with the `State` and `Event` types that can probably not be inferred or checked.

Reply:
Another idea could be that we allow them to be combined, and if both are defined it means that `snapshotWhen` is like extra snapshotting at specific points that make sense for the business, but it's also snapshotting with the interval defined by numberOfEvents in RetentionCriteria, and it's only those that trigger deletes. Makes it a bit more complicated in the implementation but might make sense for user?

Reply:
I like that second idea with extra snapshots, but, wouldn't it be nice also to be able to only have that possibility though, at specific state transitions.

Reply:
I agree, it's the right way.

It will be possible to use snapshotWhen without retention. The part that will be missing then is deletes of those snapshots (since deletes are only triggered by retention criteria). I think that is ok. We can defer that to https://github.com/akka/akka/issues/26550 or some other future retention criteria.